File: Uni_MS/angular-ui/src/app/models/invoice.ts
import { Student } from './student';
import { Semester } from './semester';
import { FeeType } from './fee-type';

export interface Invoice {
  id?: number;
  invoiceNumber?: string;
  studentId?: number;
  student?: Student;
  academicYear?: string;
  semesterId?: number;
  semester?: Semester;
  totalAmount?: number;
  paidAmount?: number;
  dueAmount?: number;
  discountAmount?: number;
  fineAmount?: number;
  status?: string;
  dueDate?: string;
  notes?: string;
  items?: InvoiceItem[];
  createdAt?: string;
  updatedAt?: string;
}

export interface InvoiceItem {
  id?: number;
  invoiceId?: number;
  feeTypeId?: number;
  feeType?: FeeType;
  description?: string;
  amount?: number;
  discountAmount?: number;
  netAmount?: number;
}

File: Uni_MS/uni_ms/src/main/java/com/badrulamin/University_Management/entity/Invoice.java
package com.badrulamin.University_Management.entity;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "invoices")
public class Invoice extends BaseEntity {

    @NotBlank
    @Size(max = 30)
    @Column(name = "invoice_number", unique = true, nullable = false)
    private String invoiceNumber;

    @NotNull
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @Size(max = 20)
    @Column(name = "academic_year")
    private String academicYear;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "semester_id")
    private Semester semester;

    @Column(name = "total_amount")
    private Double totalAmount = 0.0;

    @Column(name = "paid_amount")
    private Double paidAmount = 0.0;

    @Column(name = "due_amount")
    private Double dueAmount = 0.0;

    @Column(name = "discount_amount")
    private Double discountAmount = 0.0;

    @Column(name = "fine_amount")
    private Double fineAmount = 0.0;

    @Size(max = 20)
    @Column(nullable = false)
    private String status = "PENDING";

    private LocalDate dueDate;

    private String notes;

    @JsonProperty("studentId")
    public void setStudentId(Long id) {
        if (id != null) {
            this.student = new Student();
            this.student.setId(id);
        }
    }

    @JsonProperty
    public Long getStudentId() {
        return this.student != null ? this.student.getId() : null;
    }

    @JsonProperty("semesterId")
    public void setSemesterId(Long id) {
        if (id != null) {
            this.semester = new Semester();
            this.semester.setId(id);
        }
    }

    @JsonProperty
    public Long getSemesterId() {
        return this.semester != null ? this.semester.getId() : null;
    }
}
